elasticdl/python/allreduce/tensorflow_optimizer.py

```diff
@@ -0,0 +1,590 @@
+            assert len(self.locally_aggregated_grads) + \
+                self.num_none_grad_updates == len(grads)
+
+        # We expect to get a `sess` when we need to manually do a `sess.run(...)`
+        # for the variables to be initialized. This is the `tf.keras`
+        # optimizers.
+        if self.optimizer_type == self._OPTIMIZER_TYPE_KERAS:
+            session = tf.compat.v1.keras.backend.get_session(op_input_list=())
+            vars_init_op = tf.compat.v1.variables_initializer(
+                [self.counter, *get_not_none_from_list(self.locally_aggregated_grads)]
+            )
+            session.run(vars_init_op)
+
+    def _clear_grads(self):
+        clear_ops_list = []
+        for idx, grad_aggregator in enumerate(self.locally_aggregated_grads):
+            clear_op = grad_aggregator.assign(grad_aggregator.initial_value)
+            clear_ops_list.append(clear_op)
+        return tf.group(*clear_ops_list)
+
+    def _aggregate_grads(self, grads):
+        aggregation_ops_list = []
+        grads = get_not_none_from_list(grads)
+        assert len(grads) == len(self.locally_aggregated_grads)
+
+        # Apply new gradient updates to the local copy.
+        for idx, grad in enumerate(grads):
+            if self.sparse_as_dense and isinstance(grad, tf.IndexedSlices):
+                grad = tf.convert_to_tensor(grad)
+
+            updated_grad_aggregator = self.locally_aggregated_grads[idx].assign_add(
+                grad)
+            aggregation_ops_list.append(updated_grad_aggregator)
+
+        return aggregation_ops_list
+
+    def _allreduce_grads_helper(self, grads):
+        # Read in latest variables values.
+        aggregated_grads = []
+        aggregation_read_ops_list = []
+        for idx, locally_aggregated_grad in enumerate(
+                self.locally_aggregated_grads):
+            aggregated_grads.append(locally_aggregated_grad.read_value())
+            aggregation_read_ops_list.append(aggregated_grads[idx])
+        aggregation_read_ops = tf.group(*aggregation_read_ops_list)
+
+        with tf.control_dependencies([aggregation_read_ops]):
+            averaged_gradients = self._allreduce_grads(aggregated_grads)
+
+            # Reset counter.
+            with tf.control_dependencies([g.op for g in averaged_gradients if g is not None]):
+                reset_op = self.counter.assign(
+                    tf.constant(0), use_locking=True)
+
+            # *Elastic Update*: If ReduceOP is average,
+            # multiply horovod_size / global_batch_count_per_step
+            with tf.control_dependencies([reset_op]):
+                averaged_gradients = self.update_gradients_for_elastic_workers(averaged_gradients)
+
+                # Divide by backward_passes_per_step if
+                # average_aggregated_gradients is True.
+                with tf.control_dependencies([g.op for g in averaged_gradients if g is not None]):
+                # with tf.control_dependencies([reset_op]):
```

Comment:
Delete the commented code if it is unused.

Reply:
Good catch, updated.